modules/services/agentless-scanning/main.tf

```diff
@@ -0,0 +1,245 @@
+    # string Sysdig.
+    condition {
+      test     = "StringEquals"
+      variable = "aws:RequestTag/CreatedBy"
+      values   = ["Sysdig"]
+    }
+  }
+
+  # This statement allows the modification of those snapshot that have
+  # a simple "CreatedBy" tag valued "Sysdig". Additionally, such
+  # snapshots can only be shared with a specific AWS account, namely
+  # Sysdig account.
+  statement {
+    sid = "ec2SnapshotShare"
+
+    actions = [
+      "ec2:ModifySnapshotAttribute",
+    ]
+
+    condition {
+      test     = "StringEqualsIgnoreCase"
+      variable = "aws:ResourceTag/CreatedBy"
+      values   = ["Sysdig"]
+    }
+
+    condition {
+      test     = "StringEquals"
+      variable = "ec2:Add/userId"
+      values = [
+        local.agentless_account_id
+      ]
+    }
+
+    resources = [
+      "*",
+    ]
+  }
+
+  statement {
+    sid = "ec2SnapshotDelete"
+
+    actions = [
+      "ec2:DeleteSnapshot",
+    ]
+
+    condition {
+      test     = "StringEqualsIgnoreCase"
+      variable = "aws:ResourceTag/CreatedBy"
+      values   = ["Sysdig"]
+    }
+
+    resources = [
+      "*",
+    ]
+  }
+}
+
+resource "aws_iam_policy" "agentless" {
+  name        = "sysdig-secure-agentless-assume-role"
+  path        = "/sysdig/secure/agentless/"
+  description = "Grants Sysdig Secure access to volumes and snapshots"
+  policy      = data.aws_iam_policy_document.agentless.json
+}
+
+data "aws_iam_policy_document" "agentless_assume_role_policy" {
+  statement {
+    sid = "SysdigSecureAgentless"
+
+    actions = [
+      "sts:AssumeRole"
+    ]
+
+    principals {
+      type = "AWS"
+      identifiers = [
+        var.trusted_identity,
+      ]
+    }
+
+    condition {
+      test     = "StringEquals"
+      variable = "sts:ExternalId"
+      values   = [var.external_id]
+    }
+  }
+}
+
+resource "aws_iam_role" "agentless" {
+  name               = var.name
+  tags               = var.tags
+  assume_role_policy = data.aws_iam_policy_document.agentless_assume_role_policy.json
+}
+
+resource "aws_iam_policy_attachment" "agentless" {
+  name       = "sysdig-agentless-host-scanning"
+  roles      = [aws_iam_role.agentless.name]
+  policy_arn = aws_iam_policy.agentless.arn
+}
+
+data "aws_iam_policy_document" "key_policy" {
+  statement {
+    sid = "SysdigAllowKms"
+
+    principals {
+      type = "AWS"
+      identifiers = [
+        "arn:aws:iam::${local.agentless_account_id}:root",
+        var.trusted_identity,
+        aws_iam_role.agentless.arn,
+      ]
+    }
+
+    actions = [
+      "kms:Encrypt",
+      "kms:Decrypt",
+      "kms:ReEncrypt*",
+      "kms:GenerateDataKey*",
+      "kms:DescribeKey",
+      "kms:CreateGrant",
+      "kms:ListGrants",
+    ]
+
+    resources = [
+      "*"
+    ]
+  }
+
+  statement {
+    sid = "AllowCustomerManagement"
+
+    principals {
+      type = "AWS"
+      identifiers = [
+        "arn:aws:iam::${local.account_id}:root",
+        local.caller_arn,
```

Comment:
It looks like it, but I think it'd be difficult to make the case that this is a good and desirable practice. 

@blkt @acaprari

Reply:
Ok, so the problem there is that, in AWS, you may end up creating a KMS Key that cannot be managed at all (even by the root account) if you mess up the principals.

As a solution, we added both the root account and `localcaller_arn` because it is probably a broader set of people than those who can access the root account.

In other words, if you can "run terraform" you should also be entitled to manage the key.

Reply:
@blkt is this a general limitation that somehow unique to KMS keys? The strange scenario you describes sounds like, an AWS user with permissions enough to _create_ a KMS key, is unable to then, after having created it, in turn delete it, use it, or otherwise perform other API actions on it.  Am I understanding this correctly? If so, is there any documentation that describes this uniqueness?

Part of my hesitation on this is that the peculiarities of your observed may actually be the product of, or influenced by sysdig internal permission policies, rather than some native idiosyncrasy of KMS keys in general. Another part of the hesitation around this is it somewhat breaks idempotency. With this change, the permission scheme of the installation becomes a function of _who_ did the installing, and is likely to change over the lifecycle of feature consumption.

Can you assert what happens if this line is removed?  Is the root user then no longer able to administer the key despite it's enumeration on line 218? Also, why is it that that the `agentless` role is not called out specifically here. Isn't that the primary intended usage pattern?

Reply:
Here's the [link](https://docs.aws.amazon.com/kms/latest/developerguide/key-policies.html) to the documentation where you can read

> No AWS principal, including the account root user or key creator, has any permissions to a KMS key unless they are explicitly allowed, and never denied, in a key policy, IAM policy, or grant.

This [docs](https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-default.html#key-policy-default-allow-root-enable-iam) state this more clearly as well

> Unlike other AWS resource policies, an AWS KMS key policy does not automatically give permission to the account or any of its principals. To give permission to any principal, including the [account principal](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-accounts), you must use a key policy statement that provides the permission explicitly. You are not required to give the account principal, or any principal, access to the KMS key. However, giving access to the account principal helps you prevent the key from becoming unmanageable.

As long as `arn:aws:iam::${local.account_id}:root` is there it should be fine, problem is that any `terraform destroy` would be impossible to run if not using the root account, which is unlikely to happen. For example our INFRA team assumes a role with administrator-level permissions to run Terraform.

Finally, `agentless` role does not technically need direct access or manage rights on the Key, it just needs an indirect permission to use it during the `CopySnapshot` procedure which involves creating grants, encryption, decryption and re-encryption, which is granted at line 193.

My suggestion would be to at least allow the Customer to specify an ARN additional to the root account, so that their INFRA team can manage it.

Reply:
understood. thanks for the detailed explanation.  I think we may want to parameterize this in the future to allow meaningful overrides for the user, but for now, this seems MVP for me 👍